File: common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/Work.scala
package uk.ac.wellcome.models.work.internal

import java.time.Instant

/** Work is the core model in the pipeline / API.
  *
  * It is parameterised by State, meaning the same type of Work can be in a
  * number of possible states depending on where in the pipeline it is.
  */
sealed trait Work[State <: WorkState] {

  val version: Int
  val state: State
  val data: WorkData[State#WorkDataState]

  def sourceIdentifier: SourceIdentifier = state.sourceIdentifier

  def id: String = state.id

  def identifiers: List[SourceIdentifier] =
    sourceIdentifier :: data.otherIdentifiers

  def transition[OutState <: WorkState](args: OutState#TransitionArgs = ())(
    implicit transition: WorkFsm.Transition[State, OutState])
    : Work[OutState] = {
    val outState = transition.state(state, data, args)
    val outData = transition.data(data)
    this match {
      case Work.Visible(version, _, _) =>
        Work.Visible(version, outData, outState)
      case Work.Invisible(version, _, _, invisibilityReasons) =>
        Work.Invisible(version, outData, outState, invisibilityReasons)
      case Work.Deleted(version, _, _, deletedReason) =>
        Work.Deleted(version, outData, outState, deletedReason)
      case Work.Redirected(version, redirect, _) =>
        Work.Redirected(version, transition.redirect(redirect), outState)
    }
  }
}

object Work {

  case class Visible[State <: WorkState](
    version: Int,
    data: WorkData[State#WorkDataState],
    state: State,
  ) extends Work[State]

  case class Redirected[State <: WorkState](
    version: Int,
    redirect: State#WorkDataState#Id,
    state: State,
  ) extends Work[State] {
    val data = WorkData[State#WorkDataState]()
  }

  case class Invisible[State <: WorkState](
    version: Int,
    data: WorkData[State#WorkDataState],
    state: State,
    invisibilityReasons: List[InvisibilityReason] = Nil,
  ) extends Work[State]

  case class Deleted[State <: WorkState](
    version: Int,
    data: WorkData[State#WorkDataState],
    state: State,
    deletedReason: DeletedReason,
  ) extends Work[State]
}

/** WorkData contains data common to all types of works that can exist at any
  * stage of the pipeline.
  */
case class WorkData[State <: DataState](
  title: Option[String] = None,
  otherIdentifiers: List[SourceIdentifier] = Nil,
  mergeCandidates: List[MergeCandidate[State#Id]] = Nil,
  alternativeTitles: List[String] = Nil,
  format: Option[Format] = None,
  description: Option[String] = None,
  physicalDescription: Option[String] = None,
  lettering: Option[String] = None,
  createdDate: Option[Period[State#MaybeId]] = None,
  subjects: List[Subject[State#MaybeId]] = Nil,
  genres: List[Genre[State#MaybeId]] = Nil,
  contributors: List[Contributor[State#MaybeId]] = Nil,
  thumbnail: Option[Location] = None,
  production: List[ProductionEvent[State#MaybeId]] = Nil,
  languages: List[Language] = Nil,
  edition: Option[String] = None,
  notes: List[Note] = Nil,
  duration: Option[Int] = None,
  items: List[Item[State#MaybeId]] = Nil,
  collectionPath: Option[CollectionPath] = None,
  imageData: List[ImageData[State#Id]] = Nil,
  workType: WorkType = WorkType.Standard,
)

/** WorkState represents the state of the work in the pipeline, and contains
  * different data depending on what state it is. This allows us to consider the
  * Work model as a finite state machine with the following stages corresponding
  * to stages of the pipeline:
  *
  *      |
  *      | (transformer)
  *      ▼
  *    Source
  *      |
  *      | (id minter)
  *      ▼
  *   Identified
  *      |
  *      | (matcher / merger)
  *      ▼
  *  Merged
  *      |
  *      | (relation embedder)
  *      ▼
  * Denormalised
  *      |
  *      | (ingestor)
  *      ▼
  *   Indexed
  */
sealed trait WorkState {

  type WorkDataState <: DataState
  type TransitionArgs

  val sourceIdentifier: SourceIdentifier
  val modifiedTime: Instant
  val relations: Relations
  def id: String
}

object WorkState {

  case class Source(
    sourceIdentifier: SourceIdentifier,
    modifiedTime: Instant
  ) extends WorkState {

    type WorkDataState = DataState.Unidentified
    type TransitionArgs = Unit

    def id = sourceIdentifier.toString
    val relations = Relations.none
  }

  case class Identified(
    sourceIdentifier: SourceIdentifier,
    canonicalId: String,
    modifiedTime: Instant,
  ) extends WorkState {

    type WorkDataState = DataState.Identified
    type TransitionArgs = Unit

    def id = canonicalId
    val relations = Relations.none
  }

  case class Merged(
    sourceIdentifier: SourceIdentifier,
    canonicalId: String,
    modifiedTime: Instant,
  ) extends WorkState {

    type WorkDataState = DataState.Identified
    type TransitionArgs = Instant

    def id: String = canonicalId
    val relations: Relations = Relations.none
  }

  case class Denormalised(
    sourceIdentifier: SourceIdentifier,
    canonicalId: String,
    modifiedTime: Instant,
    relations: Relations = Relations.none
  ) extends WorkState {

    type WorkDataState = DataState.Identified
    type TransitionArgs = Relations

    def id = canonicalId
  }

  case class Indexed(
    sourceIdentifier: SourceIdentifier,
    canonicalId: String,
    modifiedTime: Instant,
    derivedData: DerivedWorkData,
    relations: Relations = Relations.none
  ) extends WorkState {

    type WorkDataState = DataState.Identified
    type TransitionArgs = Unit

    def id = canonicalId
  }
}

/** The WorkFsm contains all possible transitions between work states.
  *
  * The `transition` method on `Work` allows invocation of these transitions by
  * providing the type parameter of the new state and it's respective arguments.
  */
object WorkFsm {

  import WorkState._

  sealed trait Transition[InState <: WorkState, OutState <: WorkState] {

    def state(state: InState,
              data: WorkData[InState#WorkDataState],
              args: OutState#TransitionArgs): OutState

    def data(
      data: WorkData[InState#WorkDataState]): WorkData[OutState#WorkDataState]

    def redirect(redirect: InState#WorkDataState#Id): OutState#WorkDataState#Id
  }

  implicit val identifiedToMerged = new Transition[Identified, Merged] {
    def state(state: Identified,
              data: WorkData[DataState.Identified],
              modifiedTime: Instant): Merged =
      Merged(
        sourceIdentifier = state.sourceIdentifier,
        canonicalId = state.id,
        modifiedTime = modifiedTime
      )

    def data(data: WorkData[DataState.Identified]) = data

    def redirect(redirect: IdState.Identified): IdState.Identified = redirect
  }

  implicit val mergedToDenormalised =
    new Transition[Merged, Denormalised] {
      def state(state: Merged,
                data: WorkData[DataState.Identified],
                relations: Relations): Denormalised =
        Denormalised(
          sourceIdentifier = state.sourceIdentifier,
          canonicalId = state.canonicalId,
          modifiedTime = state.modifiedTime,
          relations = relations
        )

      def data(data: WorkData[DataState.Identified]) = data

      def redirect(redirect: IdState.Identified) = redirect
    }

  implicit val denormalisedToIndexed = new Transition[Denormalised, Indexed] {
    def state(state: Denormalised,
              data: WorkData[DataState.Identified],
              args: Unit = ()): Indexed =
      Indexed(
        sourceIdentifier = state.sourceIdentifier,
        canonicalId = state.canonicalId,
        modifiedTime = state.modifiedTime,
        derivedData = DerivedWorkData(data),
        relations = state.relations
      )

    def data(data: WorkData[DataState.Identified]) = data

    def redirect(redirect: IdState.Identified) = redirect
  }
}

File: catalogue_graph/src/ingestor/models/display/work.py
from ingestor.extractors.works_extractor import VisibleExtractedWork
from ingestor.models.shared.serialisable import ElasticsearchModel
from ingestor.transformers.work_display_transformer import DisplayWorkTransformer
from utils.types import DisplayWorkType

from .concept import DisplayConcept, DisplayContributor, DisplayGenre, DisplaySubject
from .holdings import DisplayHoldings
from .id_label import DisplayId, DisplayIdLabel
from .identifier import DisplayIdentifier
from .item import DisplayItem
from .location import DisplayDigitalLocation
from .note import DisplayNote
from .production_event import DisplayProductionEvent
from .relation import DisplayRelation


class DisplayWork(ElasticsearchModel):
    id: str
    title: str | None
    alternative_titles: list[str]
    reference_number: str | None
    description: str | None
    physical_description: str | None
    work_type: DisplayIdLabel | None
    lettering: str | None
    created_date: DisplayConcept | None
    contributors: list[DisplayContributor]
    identifiers: list[DisplayIdentifier]
    subjects: list[DisplaySubject]
    genres: list[DisplayGenre]
    thumbnail: DisplayDigitalLocation | None
    items: list[DisplayItem]
    holdings: list[DisplayHoldings]
    availabilities: list[DisplayIdLabel]
    production: list[DisplayProductionEvent]
    languages: list[DisplayIdLabel]
    edition: str | None
    notes: list[DisplayNote]
    duration: int | None
    current_frequency: str | None
    former_frequency: list[str]
    designation: list[str]
    images: list[DisplayId]
    parts: list[DisplayRelation]
    part_of: list[DisplayRelation]
    type: DisplayWorkType

    @classmethod
    def from_extracted_work(cls, extracted: VisibleExtractedWork) -> "DisplayWork":
        work = extracted.work
        transformer = DisplayWorkTransformer(extracted)

        return DisplayWork(
            id=work.state.canonical_id,
            title=work.data.title,
            alternative_titles=work.data.alternative_titles,
            reference_number=work.data.reference_number,
            description=work.data.description,
            physical_description=work.data.physical_description,
            work_type=transformer.work_type,
            lettering=work.data.lettering,
            created_date=transformer.created_date,
            thumbnail=transformer.thumbnail,
            items=list(transformer.items),
            holdings=list(transformer.holdings),
            production=list(transformer.production),
            languages=list(transformer.languages),
            edition=work.data.edition,
            notes=list(transformer.notes),
            duration=work.data.duration,
            current_frequency=work.data.current_frequency,
            former_frequency=work.data.former_frequency,
            designation=work.data.designation,
            images=list(transformer.images),
            identifiers=list(transformer.identifiers),
            contributors=list(transformer.contributors),
            genres=list(transformer.genres),
            subjects=list(transformer.subjects),
            availabilities=transformer.availabilities,
            parts=list(transformer.parts),
            part_of=list(transformer.part_of),
            type=work.data.display_work_type,
        )
